pkg/server/io/logger.go

```diff
@@ -51,46 +46,103 @@ func NewDiscardLogger() io.WriteCloser {
 	defer rc.Close()
-	defer wc.Close()
-	streamBytes := []byte(stream)
-	delimiterBytes := []byte{delimiter}
-	partialBytes := []byte(runtime.LogTagPartial)
-	fullBytes := []byte(runtime.LogTagFull)
+	var (
+		stream    = []byte(s)
+		delimiter = []byte{delimiter}
+		partial   = []byte(runtime.LogTagPartial)
+		full      = []byte(runtime.LogTagFull)
+		buf       [][]byte
+		length    int
+		bufSize   = bufSize
+	)
+	if maxLen > 0 && maxLen < bufSize {
+		bufSize = maxLen
+	}
 	r := bufio.NewReaderSize(rc, bufSize)
 	for {
-		lineBytes, isPrefix, err := r.ReadLine()
-		if err == io.EOF {
-			logrus.Debugf("Finish redirecting log file %q", path)
-			return
-		}
+		var stop bool
+		newLine, isPrefix, err := r.ReadLine()
 		if err != nil {
-			logrus.WithError(err).Errorf("An error occurred when redirecting log file %q", path)
-			return
+			if err == io.EOF {
+				logrus.Debugf("Getting EOF from stream %q while redirecting to log file %q", s, path)
+			} else {
+				logrus.WithError(err).Errorf("An error occurred when redirecting stream %q to log file %q", s, path)
+			}
+			if length == 0 {
+				// No content left to write, break.
+				break
+			}
+			// Stop after writing the content left in buffer.
+			stop = true
+		} else {
+			// Buffer returned by ReadLine will change after
+			// next read, copy it.
+			l := make([]byte, len(newLine))
+			copy(l, newLine)
+			buf = append(buf, l)
+			length += len(l)
 		}
-		tagBytes := fullBytes
-		if isPrefix {
-			tagBytes = partialBytes
+		var tag, line []byte
+		switch {
+		case maxLen < 0:
+			// Always try reading the full line, if no max length limit.
+			fallthrough
+		case length < maxLen:
+			if isPrefix {
+				continue
+			}
+			tag = full
+			line = bytes.Join(buf, nil)
+			buf = nil
+			length = 0
+		case length > maxLen:
+			tag = partial
+			last := buf[len(buf)-1]
+			exceedLen := length - maxLen
+			// len(last) must be larger than exceedLen, or else
+			// the buffer should have be written in the previous
+			// iteration.
+			buf[len(buf)-1] = last[:len(last)-exceedLen]
+			line = bytes.Join(buf, nil)
+			buf = [][]byte{last[len(last)-exceedLen:]}
+			length = exceedLen
+		case length == maxLen:
+			if isPrefix {
+				tag = partial
+			} else {
+				tag = full
+			}
+			line = bytes.Join(buf, nil)
+			buf = nil
+			length = 0
 		}
-		timestampBytes := time.Now().AppendFormat(nil, time.RFC3339Nano)
-		data := bytes.Join([][]byte{timestampBytes, streamBytes, tagBytes, lineBytes}, delimiterBytes)
+		timestamp := time.Now().AppendFormat(nil, timestampFormat)
+		data := bytes.Join([][]byte{timestamp, stream, tag, line}, delimiter)
 		data = append(data, eol)
-		if _, err := wc.Write(data); err != nil {
-			logrus.WithError(err).Errorf("Fail to write %q log to log file %q", stream, path)
+		if _, err := w.Write(data); err != nil {
+			logrus.WithError(err).Errorf("Fail to write %q log to log file %q", s, path)
 		}
 		// Continue on write error to drain the input.
+		if stop {
```

Comment:
We've already drain the input. I just mean we should not return error if we fail to write one line, because we should continue draining the container output.

I put the comment in a wrong place. Updated.